Sources/Core/Sources/Server/Ping/Pinger.swift

```diff
@@ -29,46 +28,47 @@ public class Pinger: ObservableObject {
     switch event {
       case let event as ConnectionFailedEvent:
         ThreadUtil.runInMain {
-          response = Result.failure(PingError.connectionFailed(event.networkError))
+          self.response = Result.failure(PingError.connectionFailed(event.networkError))
         }
       default:
         break
     }
   }
 
-  public func ping() throws {
+  public func ping() async throws {
     if let connection = connection {
       ThreadUtil.runInMain {
-        response = nil
+        self.response = nil
       }
       try connection.ping()
       shouldPing = false
     } else if !isConnecting {
       shouldPing = true
-      connect()
+      await connect()
     } else {
       shouldPing = true
     }
   }
 
-  private func connect() {
+  private func connect() async {
     isConnecting = true
-    // DNS resolution sometimes takes a while so we do that in parallel
-    queue.async {
-      do {
-        let connection = try ServerConnection(descriptor: self.descriptor)
-        connection.setPacketHandler(self.handlePacket)
-        connection.eventBus.registerHandler(self.handleNetworkEvent)
-        self.connection = connection
-        self.isConnecting = false
-        if self.shouldPing {
-          try? self.ping()
-        }
-      } catch {
-        self.isConnecting = false
-        log.trace("Failed to create server connection")
-        ThreadUtil.runInMain {
-          self.response = Result.failure(.connectionFailed(error))
+    await withTaskGroup(of: Void.self) { group in
+      group.addTask {
```

Comment:
I think you can ditch these two lines (and the corresponding closing braces) entirely, the wonders of structured concurrency 😎 The reason is that you call connect from a `Task` so it's already kinda in the background, the task group would just be jumping to a new task with no real effect